heartbeat/service.go

```diff
@@ -0,0 +1,172 @@
+	if ch.IsZero() {
+		return nil
+	}
+
+	var found []account.ParticipationRecordForRound
+	for _, pr := range s.accts.Keys(current + 1) { // only look at accounts we have part keys for
+		acct, _, _, err := s.ledger.LookupAccount(current, pr.Account)
+		if err != nil {
+			s.log.Errorf("error looking up %v: %v", pr.Account, err)
+			continue
+		}
+		if acct.Status == basics.Online {
+			if ch.Failed(pr.Account, acct.LastSeen()) {
+				s.log.Infof(" %v needs a heartbeat\n", pr.Account)
+				found = append(found, pr)
+			}
+		}
+		/* If we add a grace period to suspension for absenteeism, then we could
+		   also make it free to heartbeat during that period. */
+	}
+	return found
+}
+
+// loop monitors for any of Service's participants being suspended. If they are,
+// it tries to being them back online by emitting a heartbeat transaction. It
+// could try to predict an upcoming suspension, which would prevent the
+// suspension from ever occurring, but that would be considerably more complex
+// both to avoid emitting repeated heartbeats, and to ensure the prediction and
+// the suspension logic match.  This feels like a cleaner end-to-end test, at
+// the cost of lost couple rounds of participation. (Though suspension is
+// designed to be extremely unlikely anyway.)
+func (s *Service) loop() {
+	defer s.wg.Done()
+	latest := s.ledger.LastRound()
+	for {
+		// exit if Done, else wait for next round
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-s.ledger.WaitMem(latest + 1):
+		}
+
+		latest = s.ledger.LastRound()
+
+		lastHdr, err := s.ledger.BlockHdr(latest)
+		if err != nil {
+			s.log.Errorf("heartbeat service could not fetch block header for round %d: %v", latest, err)
+			continue // Try again next round, I guess?
+		}
+		proto := config.Consensus[lastHdr.CurrentProtocol]
+
+		for _, pr := range s.findChallenged(proto.Payouts) {
+			stxn := s.prepareHeartbeat(pr, lastHdr)
+			s.log.Infof("sending heartbeat %v for %v\n", stxn.Txn.HeartbeatTxnFields, pr.Account)
+			err = s.bcast.BroadcastInternalSignedTxGroup([]transactions.SignedTxn{stxn})
+			if err != nil {
+				s.log.Errorf("error broadcasting heartbeat %v for %v: %v", stxn, pr.Account, err)
+			}
+		}
+	}
+}
+
+// acceptingByteCode is the byte code to a logic signature that will accept anything (except rekeying).
+var acceptingByteCode = logic.MustAssemble(`
+#pragma version 11
+txn RekeyTo; global ZeroAddress; ==
+`)
+var acceptingSender = basics.Address(logic.HashProgram(acceptingByteCode))
+
+func (s *Service) prepareHeartbeat(pr account.ParticipationRecordForRound, latest bookkeeping.BlockHeader) transactions.SignedTxn {
+	var stxn transactions.SignedTxn
+	stxn.Lsig = transactions.LogicSig{Logic: acceptingByteCode}
+	stxn.Txn.Type = protocol.HeartbeatTx
+	stxn.Txn.Header = transactions.Header{
+		Sender:      acceptingSender,
+		FirstValid:  latest.Round + 1,
+		LastValid:   latest.Round + 1 + 100, // maybe use the grace period?
```

Comment:
I think I'll make it a constant, smaller, and suppress further heartbeats until it passes.  Right now, the service will try to heartbeat every round starting at `grace/2`. But really, once it sends the txn, it out to relax and not make another heartbeat until after lastValid (by which time, hopefully, it will notice it's not under challenge any more).